### PR TITLE
Adding Python notebook for native function examples

### PR DESCRIPTION
**Pull request by carterwilliamson**
### Motivation and Context
This PR is intended to give some functioning examples to help users get started with python native functions.
Currently, there are no examples or very limited examples. When I was playing around with this tool I ended up using some of the integration tests as examples to help get started. Hopefully, this will help the next user with the lessons learned.

### Description
This adds a jupyter notebook that walks users through using native functions alongside semantic functions to build a simple story writing app. This notebook also shows how to use ContextVariables and run a semantic function from within a native function.

### Contribution Checklist
- [ ] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [ ] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [ ] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by carterwilliamson** [comment]
@microsoft-github-policy-service agree

**Comment by carterwilliamson** [comment]
> @carterwilliamson please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

**Comment by alexchaomander** [comment]
Thank you so much for creating this! I was just putting on our backlog the need to have better docs/notebooks for this. I'll review!

**Comment by carterwilliamson** [comment]
Happy to help out! This project is fascinating, I'm excited to be able to contribute.

File: samples/notebooks/python/07-native-function-inline.ipynb [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+    },
+    {
+      "attachments": {},
+      "cell_type": "markdown",
+      "id": "40201641",
+      "metadata": {},
+      "source": [
+        "Two of the previous notebooks showed how to [execute semantic functions inline](./03-running-semantic-function-inline.ipynb) and how to [run prompts from a file](./02-running-prompts-from-file.ipynb).\n",
+        "\n",
+        "In this notebook, we'll show how to use native functions from a file. We will also show how to call semantic functions from native functions.\n",
```
Comment (alexchaomander) [inline samples/notebooks/python/07-native-function-inline.ipynb:26]: So in addition to this, one thing that we need to do a better job showing is how to within the prompt of a semantic function, how we can mix and call other native and semantic functions from various skills. 

Right now, the only example showing this is memory notebook calling the TextMemory skill's recall function and it's not the most obvious to people that that's actually a native function being called inside this new semantic function. 

Could you extend the examples you have here to spell that out more? Let me know if you need more detail!

Comment (alexchaomander) [reply samples/notebooks/python/07-native-function-inline.ipynb:26]: Making use of some of the Semantic Kernel's already built-in core skills could be good.

Comment (alexchaomander) [reply samples/notebooks/python/07-native-function-inline.ipynb:26]: And also highlighting and explaining the `@sk_function_context_parameter` decorator would be great too

Comment (alexchaomander) [reply samples/notebooks/python/07-native-function-inline.ipynb:26]: So overall in the notebooks, we want to take a reader through the journey of:

1) How to create semantic skills with prompts. Introduce the prompt templating language
2) How to use context variables to populate values into the prompt
3) How to mix and call Semantic functions and Native Functions within the prompt (this notebook)

Comment (carterwilliamson) [reply samples/notebooks/python/07-native-function-inline.ipynb:26]: @alexchaomander I've got most of those changes you requested put in. 
I could use some clarification on the `@sk_function_context_parameter` decorator though. Aside from providing documentation about what needs to be provided in the `ContextVariables`, it doesn't seem to do anything functionally. Am I missing something?

Comment (alexchaomander) [reply samples/notebooks/python/07-native-function-inline.ipynb:26]: So, it's mostly used for Planner so the Kernel will know the argument description. (See https://github.com/microsoft/semantic-kernel/pull/597)

Also from https://github.com/microsoft/semantic-kernel/issues/624, apparently in order to support functions with multiple inputs, you'd need to use the context_parameter decorator

Comment (carterwilliamson) [reply samples/notebooks/python/07-native-function-inline.ipynb:26]: Ok the interaction with the planner makes sense, but I did just test this and multiple inputs works with or without the decorator. Not sure if that should be enforced. Seems out of scope for this discussion.

I'll add info about the planner to this doc, that seems helpful.

**Review by alexchaomander — DISMISSED** [review]
Awesome! Really like this notebook. One change I would make is to rename it as `08-native-function-inline.ipynb` because we will be merging in a 07 hugging face notebook very shortly. https://github.com/microsoft/semantic-kernel/pull/762

**Comment by TMUND** [comment]
Love it ❤️

File: samples/notebooks/python/07-native-function-inline.ipynb (hunk 1 above)
Comment (carterwilliamson) [reply samples/notebooks/python/07-native-function-inline.ipynb]: @alexchaomander any other inputs/feedback or can we resolve this conversation?

File: samples/notebooks/python/08-native-function-inline.ipynb [hunk 2]
```diff
@@ -0,0 +1,685 @@
...
+      "metadata": {},
+      "source": []
+    },
+    {
+      "attachments": {},
+      "cell_type": "markdown",
+      "id": "40201641",
+      "metadata": {},
+      "source": [
+        "Two of the previous notebooks showed how to [execute semantic functions inline](./03-running-semantic-function-inline.ipynb) and how to [run prompts from a file](./02-running-prompts-from-file.ipynb).\n",
```
Comment (poweihuang0817) [inline samples/notebooks/python/08-native-function-inline.ipynb:24]: 03-running-semantic-function-inline.ipynb
There is a file name changed. Need to update here.

**Comment by poweihuang0817** [comment]
Thank you. Yes, I'm preparing a hackathon and I was spending a lot of time figuring out how to implement native functions. Thanks for the notebook

**Comment by alexchaomander** [comment]
@carterwilliamson After talking with the team, could we actually simplify the notebook to just be completely inline?

So we can remove the parts about opening an external script and writing code that you would execute inside a main function. Instead, just directly call those functions inside Jupyter! Basically, we want people to be able to do everything they need to do directly in the notebook and not have to bring up some external editor. 

Let me know if that makes sense!  We love the notebook and definitely want to bring this in!

**Comment by alexchaomander** [comment]
> @carterwilliamson After talking with the team, could we actually simplify the notebook to just be completely inline?
> 
> So we can remove the parts about opening an external script and writing code that you would execute inside a main function. Instead, just directly call those functions inside Jupyter! Basically, we want people to be able to do everything they need to do directly in the notebook and not have to bring up some external editor.
> 
> Let me know if that makes sense! We love the notebook and definitely want to bring this in!

@carterwilliamson Let me know if you're able to make the changes! Otherwise, I'd be happy to push to this branch with the updates!

**Comment by alexchaomander** [comment]
@carterwilliamson Made changes directly to this PR. Kept your examples. Just cleaned up the wording and made it more inline.

**Comment by carterwilliamson** [comment]
@alexchaomander apologies, I went on vacation and didn't have access to github. Looks like this got completed. Any remaining updates you or the team would like to see here?

**Pull request merged**